pkg/es/config/config.go

```diff
@@ -94,7 +95,8 @@ type TagsAsFields struct {
 
 // ClientBuilder creates new es.Client
 type ClientBuilder interface {
-	NewClient(logger *zap.Logger, metricsFactory metrics.Factory) (es.Client, error)
+	NewClient(logger *zap.Logger, options []elastic.ClientOptionFunc, metricsFactory metrics.Factory) (es.Client, error)
+	GetConfigOptions(logger *zap.Logger) ([]elastic.ClientOptionFunc, error)
```

Comment:
First, I just put a PR #4396 to get rid of this interface since it's an unnecessary complication. Once it merges please rebase.

Second, I don't think you need to be exposing ES options in this way. Because all other functions now just use the Configuration struct, when the password file is changed and callback is called, you can do this:
```
cfgClone = *f.primaryConfig // create a copy of the config
cfgClone.Password = // read password from file
newClient := f.newClientFn(&cfgClone, ...)
f.client.Store(newClient)
```

Reply:
> Second, I don't think you need to be exposing ES options in this way. Because all other functions now just use the Configuration struct, when the password file is changed and callback is called, you can do this:

I could have done that with the old interface too. The main reason I exposed and stored options was to store the unchangeable options. Then when password changes, we just need to append the new authentication option to the list of options. If you don't think that's necessary, I'll recreate the client from scratch as you said.